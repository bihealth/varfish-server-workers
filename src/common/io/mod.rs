//! Common, IO-related code.

pub mod std;
