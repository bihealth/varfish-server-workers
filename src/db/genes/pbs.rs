//! Data structures for (de-)serialization as generated by `prost-build`.

include!(concat!(env!("OUT_DIR"), "/varfish.v1.gene.rs"));
