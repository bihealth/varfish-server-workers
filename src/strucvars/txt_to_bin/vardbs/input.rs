//! Code supporting the I/O of public database records and a common
//! `InputRecord` for common representation.

use serde::Deserialize;
use tracing::error;

use crate::strucvars::aggregate::output::Record as InhouseDbRecord;
use crate::strucvars::query::schema::SvType;

/// dbVar database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DbVarRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// number of overall carriers
    pub num_carriers: u32,
    /// type of the SV
    pub sv_type: String,
}

/// DGV database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DgvRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural variant type
    sv_type: String,
    /// Number of observed gains.
    observed_gains: u32,
    /// Number of observed losses
    observed_losses: u32,
}

/// DGV gold standard database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DgvGsRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin_outer: i32,
    /// outer end position, 0-based
    pub end_outer: i32,
    /// The structural variant type
    pub sv_sub_type: String,
    /// Number of carriers.
    pub num_carriers: u32,
}

/// ExAC CNV database record as read from TSV file for deserialization
/// from TSV.
#[derive(Deserialize, Debug)]
pub struct ExacRecord {
    /// chromosome name
    pub chromosome: String,
    /// outer start position, 0-based
    pub begin: i32,
    /// outer end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub sv_type: String,
}

/// Thousand Genomes SV database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct G1kRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub sv_type: String,
    /// Number of hom. alt. alleles.
    pub n_homalt: u32,
    /// Number of het. alt. alleles.
    pub n_het: u32,
}

/// gnomAD SV v2 database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadSv2Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of homozygous alternative carriers
    pub n_homalt: u32,
    /// Number of heterozygous carriers
    pub n_het: u32,
}

/// gnomAD SV v4 database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadSv4Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of male homozygous reference allele carriers.
    pub male_n_homref: u32,
    /// Number of male heterozygous alternate allele carriers.
    pub male_n_het: u32,
    /// Number of male homozygous alternate allele carriers.
    pub male_n_homalt: u32,
    /// Number of male hemizygous alternate allele carriers.
    pub male_n_hemiref: u32,
    /// Number of male hemizygous reference allele carriers.
    pub male_n_hemialt: u32,
    /// Number of female homozygous reference allele carriers.
    pub female_n_homref: u32,
    /// Number of female heterozygous alternate allele carriers.
    pub female_n_het: u32,
    /// Number of female homozygous alternate allele carriers.
    pub female_n_homalt: u32,
    /// Number of samples at this site (CNV only).
    pub cnv_n_total: u32,
    /// Number of samples with a CNV at this site (CNV only).
    pub cnv_n_var: u32,
}

/// gnomAD CNV v$ database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadCnv4Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of samples at this site (passing QC).
    pub n_total: u32,
    /// Number of samples with a CNV at this site (passing QC).
    pub n_var: u32,
}

/// Common type to convert input data to.
pub struct InputRecord {
    /// Chromosome of start position.
    pub chromosome: String,
    /// Chromosome of end position.
    pub chromosome2: String,
    /// SV type
    pub sv_type: SvType,
    /// 0-based begin position
    pub begin: i32,
    /// 0-based end position
    pub end: i32,
    /// Number of carriers (or alleles), depending on database.
    pub count: u32,
}

impl TryInto<Option<InputRecord>> for InhouseDbRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome,
            chromosome2: self.chromosome2,
            sv_type: self.sv_type,
            begin: self.begin,
            end: self.end,
            count: self.carriers,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DbVarRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.split(';').next().unwrap() {
            "alu_insertion"
            | "herv_insertion"
            | "insertion"
            | "line1_insertion"
            | "mobile_element_insertion"
            | "novel_sequence_insertion"
            | "sva_insertion" => SvType::Ins,
            "copy_number_gain" | "duplication" | "tandem_duplication" => SvType::Dup,
            "alu_deletion" | "copy_number_loss" | "deletion" | "herv_deletion"
            | "line1_deletion" | "sva_deletion" => SvType::Del,
            "copy_number_variation" => SvType::Cnv,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: 1,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DgvRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_ref() {
            "alu deletion"
            | "deletion"
            | "herv deletion"
            | "line1 deletion"
            | "mobile element deletion"
            | "loss"
            | "sva deletion" => SvType::Del,
            "alu insertion"
            | "herv insertion"
            | "insertion"
            | "line1 insertion"
            | "mobile element insertion"
            | "novel sequence insertion"
            | "sva insertion" => SvType::Ins,
            "duplication" | "gain" | "tandem duplication" => SvType::Dup,
            "sequence alteration" | "complex" => return Ok(None), // skip
            "gain+loss" | "CNV" => SvType::Cnv,
            "inversion" => SvType::Inv,
            "OTHER" => return Ok(None), // skip
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: self.observed_gains + self.observed_losses,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DgvGsRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_sub_type.as_ref() {
            "Gain" => SvType::Dup,
            "Loss" => SvType::Del,
            _ => {
                error!("sv_type = {}", &self.sv_sub_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin_outer,
            end: self.end_outer,
            sv_type,
            count: self.num_carriers,
        }))
    }
}

impl TryInto<Option<InputRecord>> for ExacRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_ref() {
            "DUP" => SvType::Dup,
            "DEL" => SvType::Del,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: 1,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadSv2Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.svtype.as_str() {
            "CPX" => return Ok(None), // no correspondence
            "CTX" | "BND" => SvType::Bnd,
            "DEL" => SvType::Del,
            "DUP" => SvType::Dup,
            "INS" => SvType::Ins,
            "INV" => SvType::Inv,
            "MCNV" => SvType::Cnv,
            _ => {
                error!("sv_type = {}", &self.svtype);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin - 1,
            end: self.end,
            sv_type,
            count: self.n_homalt + self.n_het,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadCnv4Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type: match self.svtype.as_str() {
                "DEL" => SvType::Del,
                "DUP" => SvType::Dup,
                _ => {
                    error!("sv_type = {}", &self.svtype);
                    return Err("unknown SV type");
                }
            },
            count: self.n_var,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadSv4Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type: match self.svtype.as_str() {
                "BND" => SvType::Bnd,
                "CNV" => SvType::Cnv,
                "DEL" => SvType::Del,
                "DUP" => SvType::Dup,
                "INS" => SvType::Ins,
                "INV" => SvType::Inv,
                _ => {
                    error!("sv_type = {}", &self.svtype);
                    return Err("unknown SV type");
                }
            },
            count: self.male_n_het
                + self.male_n_homalt
                + self.male_n_hemialt
                + self.female_n_het
                + self.female_n_homalt
                + self.cnv_n_var,
        }))
    }
}

impl TryInto<Option<InputRecord>> for G1kRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_str() {
            "CN0" | "CNV" => SvType::Cnv,
            "DEL" => SvType::Del,
            "DEL_ALU" | "DEL_HERV" | "DEL_LINE1" | "DEL_SVA" => SvType::Del,
            "DUP" => SvType::Dup,
            "INV" => SvType::Inv,
            "INS" | "INS:ME:ALU" | "INS:ME:LINE1" | "INS:ME:SVA" => SvType::Ins,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: self.n_homalt + self.n_het,
        }))
    }
}
