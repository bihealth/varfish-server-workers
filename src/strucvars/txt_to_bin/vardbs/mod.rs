//! Code for converting other structural variant database to binary (incl. in-house).

use std::fs::File;
use std::io::Write;
use std::path::Path;
use std::time::Instant;

use anyhow::anyhow;
use prost::Message;
use thousands::Separable;

use crate::common::{build_chrom_map, trace_rss_now};
use crate::pbs::svs::{BackgroundDatabase, BgDbRecord};
use crate::strucvars::aggregate::output::Record as InhouseDbRecord;
use crate::strucvars::query::schema::SvType;

use self::input::InputRecord;

mod input;

/// Known input file types.
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum InputFileType {
    Dbvar,
    Dgv,
    DgvGs,
    Exac,
    G1k,
    GnomadSv2,
    GnomadCnv4,
    GnomadSv4,
    InhouseDb,
}
/// Deserialize from CSV reader to an `Option<records::InputRecord>`
fn deserialize_loop<Rec>(
    reader: &mut csv::Reader<Box<dyn std::io::BufRead>>,
) -> Result<Vec<BgDbRecord>, anyhow::Error>
where
    Rec: core::fmt::Debug + TryInto<Option<InputRecord>> + for<'de> serde::Deserialize<'de>,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: core::fmt::Debug,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: Send,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: std::marker::Sync,
{
    let chrom_map = build_chrom_map();
    let mut result = Vec::new();

    for record in reader.deserialize() {
        let record: Rec = record?;
        let maybe_record: Option<InputRecord> = record
            .try_into()
            .map_err(|err| anyhow!("problem with parsing: {:?}", &err))?;
        if let Some(record) = maybe_record {
            result.push(BgDbRecord {
                chrom_no: *chrom_map
                    .get(&record.chromosome)
                    .unwrap_or_else(|| panic!("unknown chrom: {:?}", &record.chromosome))
                    as i32,
                chrom_no2: *chrom_map
                    .get(&record.chromosome2)
                    .unwrap_or_else(|| panic!("unknown chrom2: {:?}", &record.chromosome2))
                    as i32,
                sv_type: match record.sv_type {
                    SvType::Del => crate::pbs::svs::SvType::Del,
                    SvType::Dup => crate::pbs::svs::SvType::Dup,
                    SvType::Inv => crate::pbs::svs::SvType::Inv,
                    SvType::Ins => crate::pbs::svs::SvType::Ins,
                    SvType::Bnd => crate::pbs::svs::SvType::Bnd,
                    SvType::Cnv => crate::pbs::svs::SvType::Cnv,
                } as i32,
                start: record.begin + 1,
                stop: record.end,
                count: record.count,
            });
        }
    }

    Ok(result)
}

/// Branch around `deserialize_loop`.
pub fn deserialize_branch(
    input_type: InputFileType,
    reader: &mut csv::Reader<Box<dyn std::io::BufRead>>,
) -> Result<Vec<BgDbRecord>, anyhow::Error> {
    match input_type {
        InputFileType::Dbvar => deserialize_loop::<input::DbVarRecord>(reader),
        InputFileType::Dgv => deserialize_loop::<input::DgvRecord>(reader),
        InputFileType::DgvGs => deserialize_loop::<input::DgvGsRecord>(reader),
        InputFileType::Exac => deserialize_loop::<input::ExacRecord>(reader),
        InputFileType::G1k => deserialize_loop::<input::G1kRecord>(reader),
        InputFileType::InhouseDb => deserialize_loop::<InhouseDbRecord>(reader),
        InputFileType::GnomadSv2 => deserialize_loop::<input::GnomadSv2Record>(reader),
        InputFileType::GnomadCnv4 => deserialize_loop::<input::GnomadCnv4Record>(reader),
        InputFileType::GnomadSv4 => deserialize_loop::<input::GnomadSv4Record>(reader),
    }
}

/// Perform conversion to protobuf `.bin` file.
pub fn convert_to_bin<P, Q>(
    path_input_tsv: P,
    path_output: Q,
    input_type: InputFileType,
) -> Result<(), anyhow::Error>
where
    P: AsRef<Path>,
    Q: AsRef<Path>,
{
    // Setup CSV reader for BED file - header is written as comment and must be
    // ignored.
    let mut reader = csv::ReaderBuilder::new()
        .has_headers(false)
        .comment(Some(b'#'))
        .delimiter(b'\t')
        .from_reader(mehari::common::io::std::open_read_maybe_gz(
            path_input_tsv.as_ref(),
        )?);
    let before_parsing = Instant::now();

    let records = deserialize_branch(input_type, &mut reader)?;
    let bg_db = BackgroundDatabase { records };

    tracing::debug!(
        "total time spent reading {} records: {:?}",
        bg_db.records.len().separate_with_commas(),
        before_parsing.elapsed()
    );
    trace_rss_now();

    let before_writing = Instant::now();
    let mut output_file = File::create(&path_output)?;
    output_file.write_all(&bg_db.encode_to_vec())?;
    output_file.sync_all()?;
    tracing::debug!(
        "total time spent writing {} records: {:?}",
        bg_db.records.len().separate_with_commas(),
        before_writing.elapsed()
    );

    Ok(())
}

#[cfg(test)]
mod test {
    use super::InputFileType;

    #[rstest::rstest]
    #[case::dbvar(
        InputFileType::Dbvar,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dbvar.bed.gz"
    )]
    #[case::dgv(
        InputFileType::Dgv,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dgv.bed.gz"
    )]
    #[case::dgv_gs(
        InputFileType::DgvGs,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dgv_gs.bed.gz"
    )]
    #[case::exac(
        InputFileType::Exac,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/exac.bed.gz"
    )]
    #[case::g1k(
        InputFileType::G1k,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/g1k.bed.gz"
    )]
    #[case::gnomad_sv2(
        InputFileType::GnomadSv2,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/gnomad_sv.bed.gz"
    )]
    #[case::gnomad_cnv4(
        InputFileType::GnomadCnv4,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch38/strucvar/gnomad-cnv.bed.gz"
    )]
    #[case::gnomad_sv4(
        InputFileType::GnomadSv4,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch38/strucvar/gnomad-sv.bed.gz"
    )]
    #[case::inhouse_db(
        InputFileType::InhouseDb,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/inhouse.tsv"
    )]
    fn test_deserialize_branch(
        #[case] input_type: InputFileType,
        #[case] path_input: &str,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!(
            "{:?}-{}",
            input_type,
            path_input
                .split('/')
                .last()
                .unwrap()
                .split('.')
                .next()
                .unwrap()
        );

        let mut reader = csv::ReaderBuilder::new()
            .has_headers(false)
            .comment(Some(b'#'))
            .delimiter(b'\t')
            .from_reader(mehari::common::io::std::open_read_maybe_gz(path_input)?);

        let records = super::deserialize_branch(input_type, &mut reader)?;
        insta::assert_yaml_snapshot!(records);

        Ok(())
    }
}
