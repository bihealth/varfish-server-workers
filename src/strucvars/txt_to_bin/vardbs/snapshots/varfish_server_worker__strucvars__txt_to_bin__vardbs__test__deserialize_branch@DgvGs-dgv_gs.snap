---
source: src/strucvars/txt_to_bin/vardbs/mod.rs
expression: records
---
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 10001
  stop: 82189
  count: 13
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 10001
  stop: 127330
  count: 30
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 10499
  stop: 235938
  count: 13
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 21064
  stop: 35983
  count: 12
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 46501
  stop: 71800
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 49298
  stop: 98602
  count: 16
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 49298
  stop: 98602
  count: 16
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 61723
  stop: 356542
  count: 37
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 70007
  stop: 88084
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 86778
  stop: 91967
  count: 13
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 87433
  stop: 89163
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 91317
  stop: 125400
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 121042
  stop: 176984
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 136181
  stop: 137811
  count: 5
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 227601
  stop: 267719
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 317720
  stop: 351915
  count: 12
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 317770
  stop: 471337
  count: 9
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 348092
  stop: 461014
  count: 3
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 350531
  stop: 435382
  count: 16
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 404112
  stop: 471337
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 437342
  stop: 468644
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 459944
  stop: 464929
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 460229
  stop: 471368
  count: 14
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 521413
  stop: 575095
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 521413
  stop: 706368
  count: 4
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 521413
  stop: 850675
  count: 6
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 521553
  stop: 546384
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 522577
  stop: 641599
  count: 3
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 544966
  stop: 569604
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 546384
  stop: 549347
  count: 13
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 563896
  stop: 570775
  count: 24
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 564452
  stop: 574333
  count: 4
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 564582
  stop: 570178
  count: 4
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 637238
  stop: 730328
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 668630
  stop: 851378
  count: 63
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 712111
  stop: 764840
  count: 202
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 713044
  stop: 755966
  count: 4
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 723901
  stop: 727245
  count: 25
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 724314
  stop: 727245
  count: 10
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 738914
  stop: 809773
  count: 20
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 751887
  stop: 762622
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 754201
  stop: 759100
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 766101
  stop: 809773
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 766408
  stop: 769112
  count: 195
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 769117
  stop: 852664
  count: 60
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 775292
  stop: 792588
  count: 16
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 812998
  stop: 1708649
  count: 3
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 834801
  stop: 1029900
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 850017
  stop: 1097335
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 869249
  stop: 870860
  count: 71
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 869310
  stop: 870645
  count: 130
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 869497
  stop: 870645
  count: 277
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 873391
  stop: 874049
  count: 8
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 947113
  stop: 948003
  count: 5
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 963659
  stop: 964567
  count: 242
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 964980
  stop: 965488
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1010973
  stop: 1014683
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1011188
  stop: 1014828
  count: 15
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1034317
  stop: 1035340
  count: 30
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1058898
  stop: 1058993
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1074379
  stop: 1076572
  count: 19
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1137003
  stop: 1138283
  count: 9
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1142700
  stop: 1143161
  count: 939
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1183198
  stop: 1183861
  count: 92
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1223189
  stop: 1226049
  count: 49
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1223580
  stop: 1225633
  count: 5
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1226063
  stop: 1314437
  count: 26
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1285180
  stop: 1287127
  count: 9
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1285340
  stop: 1286900
  count: 24
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1302378
  stop: 1302882
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1352531
  stop: 1453590
  count: 19
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1385069
  stop: 1405538
  count: 11
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1387455
  stop: 1443716
  count: 59
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1393861
  stop: 1431581
  count: 22
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1400225
  stop: 1409107
  count: 34
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1414245
  stop: 1416792
  count: 30
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1415099
  stop: 1462190
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1432368
  stop: 1470064
  count: 12
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1443538
  stop: 1445764
  count: 112
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1465911
  stop: 1466268
  count: 346
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1500941
  stop: 1672603
  count: 7
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1522874
  stop: 1524674
  count: 30
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1540850
  stop: 1541970
  count: 8
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1540930
  stop: 1541970
  count: 56
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1559466
  stop: 1683706
  count: 5
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1582548
  stop: 1647400
  count: 5
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1584529
  stop: 1685076
  count: 36
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1598413
  stop: 1598601
  count: 450
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1616990
  stop: 1653672
  count: 6
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1620684
  stop: 1692321
  count: 322
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1627805
  stop: 1692321
  count: 80
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1634170
  stop: 1655791
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1635205
  stop: 1680442
  count: 20
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 1642103
  stop: 1672603
  count: 7
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1649047
  stop: 1650677
  count: 69
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1658945
  stop: 1691900
  count: 3
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1662938
  stop: 1672603
  count: 3
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1866297
  stop: 1867172
  count: 3
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 1910347
  stop: 1911917
  count: 16
